osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+                                   {
+                                       tpBeats.Add(Math.Floor(currentTime));
+                                       currentTime += tp.BeatLength;
+                                   }
+
+                                   return tpBeats;
+                               })
+                               .Where(x => Precision.AlmostBigger(x, startTime))
+                               // Remove beats during breaks
+                               .Where(x => !beatmap.Breaks.Any(b =>
+                                   Precision.AlmostBigger(x, b.StartTime)
+                                   && Precision.AlmostBigger(origHitObjects.First(y => Precision.AlmostBigger(y.StartTime, b.EndTime)).StartTime, x)
+                               ))
+                               .ToList();
+
+            // Remove beats that are too close to the next one (e.g. due to timing point changes)
+            for (var i = beats.Count - 2; i >= 0; i--)
+            {
+                var beat = beats[i];
+
+                if (Precision.AlmostBigger(beatmap.ControlPointInfo.TimingPointAt(beat).BeatLength / 2, beats[i + 1] - beat)) beats.RemoveAt(i);
+            }
+
+            return beats;
+        }
+
+        private void addHitSamples(IEnumerable<OsuHitObject> hitObjects, IReadOnlyList<OsuHitObject> origHitObjects)
+        {
+            var lastSampleIdx = 0;
+
+            foreach (var x in hitObjects)
+            {
+                var samples = getSamplesAtTime(origHitObjects, x.StartTime);
+
+                if (samples == null)
+                {
+                    while (lastSampleIdx < origHitObjects.Count && origHitObjects[lastSampleIdx].StartTime <= x.StartTime)
+                        lastSampleIdx++;
+                    lastSampleIdx--;
+
+                    if (lastSampleIdx < 0 && lastSampleIdx >= origHitObjects.Count) continue;
+
+                    if (lastSampleIdx < origHitObjects.Count - 1)
+                    {
+                        // get samples from the next hit object if it is closer in time
+                        if (origHitObjects[lastSampleIdx + 1].StartTime - x.StartTime < x.StartTime - origHitObjects[lastSampleIdx].StartTime)
+                            lastSampleIdx++;
+                    }
+
+                    x.Samples = origHitObjects[lastSampleIdx].Samples.Where(s => !HitSampleInfo.AllAdditions.Contains(s.Name)).ToList();
+                }
+                else
+                    x.Samples = samples;
+            }
+        }
+
+        private void fixComboInfo(List<OsuHitObject> hitObjects, List<OsuHitObject> origHitObjects)
+        {
+            // First follow the combo indices in the original beatmap
+            hitObjects.ForEach(x =>
+            {
+                var origObj = origHitObjects.FindLast(y => Precision.AlmostBigger(x.StartTime, y.StartTime));
+                x.ComboIndex = origObj?.ComboIndex ?? 0;
```

Comment:
These linq expressions direly need actual explanations for what they're doing. I think this is trying to fetch the closest object preceding `x`? I can't get to that from "follow the combo indices in the original beatmap".